packages/yarnpkg-core/sources/configurationUtil.ts

```diff
@@ -0,0 +1,201 @@
+
+    if (valueType !== expectedValueType || onConflict === `hardReset`) {
+      currentResetPosition = lastRelevantPosition;
+      break;
+    }
+
+    if (valueType === ValueType.Literal)
+      return resolvedRcFile(id, value);
+
+    relevantValues.unshift([id, value]);
+
+    if (onConflict === `reset`) {
+      currentResetPosition = t;
+      break;
+    }
+
+    if (onConflict === `extend` && t === firstVisiblePosition)
+      firstVisiblePosition = 0;
+
+    lastRelevantPosition = t;
+  }
+
+  if (typeof expectedValueType === `undefined`)
+    return null;
+
+  const source = relevantValues.map(([relevantId]) => relevantId).join(`, `);
+  switch (expectedValueType) {
+    case ValueType.Array:
+      return resolvedRcFile(source, new Array<unknown>().concat(...relevantValues.map(([id, value]) => (value as Array<unknown>).map(item => attachIdToTree(item, id)))));
+
+    case ValueType.Object:{
+      const conglomerate = Object.assign({}, ...relevantValues.map(([, value]) => value));
+      const keys = Object.keys(conglomerate);
+      const result: Record<string, unknown> = {};
+
+      const nextIterationValues = rcFiles.map<[string, unknown]>(([id, data]) => {
+        return [id, getNormalized(data, key).value];
+      });
+
+      const hardResetLocation = findLastIndex(nextIterationValues, ([_, value]) => {
+        const valueType = getValueType(value);
+        return valueType !== ValueType.Object && valueType !== ValueType.Undefined;
+      });
+
+      if (hardResetLocation !== -1) {
+        const slice = nextIterationValues.slice(hardResetLocation + 1);
+        for (const key of keys) {
+          result[key] = resolveValueAt(slice, path, key, 0, slice.length);
+        }
+      } else {
+        for (const key of keys) {
+          result[key] = resolveValueAt(nextIterationValues, path, key, currentResetPosition, nextIterationValues.length);
+        }
+      }
+
+      return resolvedRcFile(source, result);
+    }
+    default:
+      throw new Error(`Assertion failed: Non-extendable value type`);
+  }
+}
+
+// Given an array of configuration files represented as tuples, which each
+// contains both an ID (for example the configuration file path) and an
+// arbitrary value, this function will traverse the whole tree to resolve
+// all `onConflict` directives.
+//
+// The returned value will recursively be turned into tuples, which each
+// contain both the ID of the configuration file that contributed the last
+// entry to the value and the final value.
+//
+export function resolveRcFiles(rcFiles: Array<[string, unknown]>) {
+  return resolveValueAt(rcFiles.map(([source, data]) => [source, {[`.`]: data}]), [], `.`, 0, rcFiles.length) as [string, Record<string, unknown>, symbol] | null;
+}
+
+export function getValue(value: unknown) {
+  return isResolvedRcFile(value) ? value[1] : value;
+}
+
+export function getSource(value: unknown) {
+  return isResolvedRcFile(value) ? value[0] : null;
+}
```

Comment:
I'm not a fan of this pattern - I'd prefer if the manipulated were always of a consistent type, even if it requires normalizing them beforehand.